lib/virtual.js

```diff
@@ -74,15 +140,104 @@ VirtualModulesPlugin.prototype.writeModule = function(filePath, contents) {
 		ctime: time,
 		birthtime: time
 	});
+	const modulePath = getModulePath(filePath, this.compiler);
 
+	// When using the WatchIgnorePlugin (https://github.com/webpack/webpack/blob/52184b897f40c75560b3630e43ca642fcac7e2cf/lib/WatchIgnorePlugin.js),
+	// the original watchFileSystem is stored in `wfs`. The following "unwraps" the ignoring
+	// wrappers, giving us access to the "real" watchFileSystem.
+	let finalWatchFileSystem = this._watcher && this._watcher.watchFileSystem;
+
+	while (finalWatchFileSystem && finalWatchFileSystem.wfs) {
+		finalWatchFileSystem = finalWatchFileSystem.wfs;
+	}
 	this.compiler.inputFileSystem._writeVirtualFile(filePath, stats, contents);
+	if (finalWatchFileSystem &&
+		(finalWatchFileSystem.watcher.fileWatchers.size ||
+			finalWatchFileSystem.watcher.fileWatchers.length)
+	) {
+		const fileWatchers = finalWatchFileSystem.watcher.fileWatchers instanceof Map ?
+			Array.from(finalWatchFileSystem.watcher.fileWatchers.values()) :
+			finalWatchFileSystem.watcher.fileWatchers;
+		fileWatchers.forEach(function(fileWatcher) {
+			if (fileWatcher.path === modulePath) {
+				delete fileWatcher.directoryWatcher._cachedTimeInfoEntries;
+				fileWatcher.directoryWatcher.setFileTime(
+					filePath,
+					time,
+					false,
+					false,
+					null
+				);
+				fileWatcher.emit("change", time, null);
+			}
+		});
+	}
 };
 
-function setData(storage, key, value) {
-	if (storage.data instanceof Map) {
-		storage.data.set(key, value);
+function getData(storage, key) {
+	// Webpack 5
+	if (storage._data instanceof Map) {
```

Comment:
you could probably simplify both this method and `setData` a lot by creating an intermediate `data` variable:
```
const data = storage._data /* webpack 5 */ || storage.data /* webpack 4 */;
if (data instanceof Map) {
  return storage.data.get(key);
}
return data.data[key];
```